Refactoring so topHits calls topHitsByStudy because it handles dupe AUROCs

leonfrench opened the pull request:
Based on my reading of the code, the current topHits function (python and R) looks for cell types from either study that have the same max AUROC value, and that's how it finds reciprocal hits. So a given reciprocal pair would have the same max AUROC and non-reciprocal pairings wouldn't.

This is probably okay in most cases, as the number of unique possible AUROC values for thousands of cells is pretty large (and cell types are large). In my case, I was getting AUROC's tied at 1, plus some of the cluster sizes are small, so it was more apparent. A simple test case is to input the same dataset twice to get ties at AUROC = 1. 

Just to be clear, the code doesn't seem to check if the two max AUROC values that are equal involve the same cell types.

For the R version, @yexilein previously rewrote the topHits function as topHitsByStudy to extract the best hits that take into account the studies. His rewrite also fixed that duplicate masking issue above, so I've changed the topHits R code to call topHitsByStudy.

leonfrench committed:
Refactoring so topHits calls topHitsByStudy because it better handles duplicate AUROC values.

yexilein reviewed (changes requested):
Hi Leon, thanks for raising the issue and offering a solution! I had not noticed this issue with the topHits function. From your description, my understanding is that it arises in the case of non-symmetrical AUROC matrices (by default, MetaNeighbor will return symmetric matrices)?

The pull request also contains changes to topHitsByStudy. Can you elaborate on these changes? I noticed that you bypass calls to getStudyId, which is not ideal for future development. I’m happy with changing the code of getStudyId itself (e.g., to make it faster or change the separator), but removing calls to getStudyId may break MetaNeighbor in the future (e.g., if we decide to change the default separator). Could you change this part of the code?

leonfrench committed:
reverting to use getStudyId

leonfrench commented:
Thank you for finding the time to review my request. 

I've reverted the code, so it calls getStudyId as before. Sorry for trading code quality for speed (I was a bit impatient). Later on, I may refactor it to handle the same study pairs before melting as that would speed it up. 

This happens with symmetrical matrices too. I haven't looked at non-symmetrical matrices, but I'm guessing it may be more of a problem for those as duplicate AUROC values probably won't mark Reciprocal_top_hits. 

Below is an example that uses the Baron data twice as input. The tophits function gives 10 Reciprocal_top_hits for the 14 cell types while topHitsByStudy returns all 14 as Reciprocal_top_hits.
```
library(MetaNeighbor)
library(scRNAseq)
library(dplyr)

#requires that you have a cached version of these datasets 
my_data <- list(
  baron = BaronPancreasData(),
  baron_duplicated = BaronPancreasData()
)

my_data$baron$"cell type" <- my_data$baron$label
my_data$baron_duplicated$"cell type" <- my_data$baron_duplicated$label
my_data$baron$"cell type" %>% unique() %>% length() #14 cell types

fused_data = mergeSCE(my_data)

global_hvgs = variableGenes(dat = fused_data, exp_labels = fused_data$study_id)

aurocs = MetaNeighborUS(var_genes = global_hvgs,
                        dat = fused_data,
                        study_id = fused_data$study_id,
                        cell_type = fused_data$"cell type",
                        fast_version = TRUE)

top_hits <- topHits(aurocs, dat = fused_data, study_id = fused_data$study_id,
        cell_type = fused_data$"cell type")
top_hits %>% group_by(Match_type) %>% count() #returns 10 Reciprocal_top_hits
top_hits %>% head(20)

topHitsByStudy(aurocs) %>% group_by(Match_type) %>% count() #returns 14 Reciprocal_top_hits

```
That example is a bit extreme, but I've noticed the same masking of reciprocal top hits in a comparison of two real datasets. In that case, I lose over 10% of the reciprocal top hits.

leonfrench committed:
Remove literal NULL legend title from heatmaps

yexilein committed:
Align parameters for topHits and topHitsByStudy

All legacy parameters are now optional (useless in new formulation, but avoids breaking old code). All novel parameters introduced by topHitsByStudy are now parameters of topHits.

yexilein commented:
Hi Leon,

thanks for making the changes rapidly and providing a reproducible example for the issue. I changed the code so that all legacy parameters of topHits become optional (it was annoying to provide all metadata as parameters, which was not really necessary) and novel parameters introduced by topHitsByStudy also become parameters for topHits.

This means that old code will still be able to run, but we are breaking the behavior of the function (to return more hits). I prefer using topHitsByStudy, so I’m happy for it to become the default behavior!
